File: kmp/composeApp/src/commonTest/kotlin/com/tanyudii/tmuxweb/domain/ServerUrlTest.kt
package com.tanyudii.tmuxweb.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/** Ports ConnectionSettingsView.swift's `url.scheme != nil && url.host != nil` validation. */
class ServerUrlTest {
    @Test
    fun `accepts http URL with host and port`() {
        assertEquals("http://192.168.1.5:5309", parseServerUrl("http://192.168.1.5:5309"))
    }

    @Test
    fun `accepts https URL`() {
        assertEquals("https://tmux.example.com", parseServerUrl("https://tmux.example.com"))
    }

    @Test
    fun `trims surrounding whitespace`() {
        assertEquals("http://host:5309", parseServerUrl("  http://host:5309  "))
    }

    @Test
    fun `drops a path suffix`() {
        assertEquals("http://host:5309", parseServerUrl("http://host:5309/some/path"))
    }

    @Test
    fun `rejects a scheme-less string`() {
        assertNull(parseServerUrl("host:5309"))
    }

    @Test
    fun `rejects a scheme with no host`() {
        assertNull(parseServerUrl("http://"))
    }

    @Test
    fun `rejects a non-http scheme`() {
        assertNull(parseServerUrl("ftp://host"))
    }

    @Test
    fun `rejects blank input`() {
        assertNull(parseServerUrl(""))
    }
}

File: web/src/domain/serverUrl.test.ts
import { describe, expect, test } from "vitest";
import { parseServerUrl } from "./serverUrl";

// Ports kmp/.../domain/ServerUrlTest.kt 1:1.
describe("parseServerUrl", () => {
  test("accepts http URL with host and port", () => {
    expect(parseServerUrl("http://192.168.1.5:5309")).toBe("http://192.168.1.5:5309");
  });

  test("accepts https URL", () => {
    expect(parseServerUrl("https://tmux.example.com")).toBe("https://tmux.example.com");
  });

  test("trims surrounding whitespace", () => {
    expect(parseServerUrl("  http://host:5309  ")).toBe("http://host:5309");
  });

  test("drops a path suffix", () => {
    expect(parseServerUrl("http://host:5309/some/path")).toBe("http://host:5309");
  });

  test("rejects a scheme-less string", () => {
    expect(parseServerUrl("host:5309")).toBeNull();
  });

  test("rejects a scheme with no host", () => {
    expect(parseServerUrl("http://")).toBeNull();
  });

  test("rejects a non-http scheme", () => {
    expect(parseServerUrl("ftp://host")).toBeNull();
  });

  test("rejects blank input", () => {
    expect(parseServerUrl("")).toBeNull();
  });
});
